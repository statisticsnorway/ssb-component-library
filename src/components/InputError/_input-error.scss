.ssb-input-error {
  @include roboto;
  background: $ssb-red-3;
  color: $ssb-white;
  display: flex;
  font-size: 14px;
  font-weight: bold;
  justify-content: center;
  line-height: normal;
  margin-top: 12px;
  padding: 4px 8px;
  position: relative;

  &.negative {
    color: $ssb-dark-5;
  }

  &:before {
    border-color: transparent transparent $ssb-red-3 transparent;
    border-style: solid;
    border-width: 8px;
    content: '';
    position: absolute;
    top: -16px;
  }
}
