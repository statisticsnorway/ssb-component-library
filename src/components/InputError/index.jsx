import React from 'react';
import PropTypes from 'prop-types';

const InputError = ({ errorMessage, negative }) => (
	<span className={`ssb-input-error ${negative ? 'negative' : ''}`}>{errorMessage}</span>
);

InputError.defaultProps = {
	negative: false,
};

InputError.propTypes = {
	errorMessage: PropTypes.string.isRequired,
	negative: PropTypes.bool,
};

export default InputError;
