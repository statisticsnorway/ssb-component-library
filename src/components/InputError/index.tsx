import React from 'react'
import classNames from '../../utils/utils'

interface InputErrorProps {
  className?: string
  errorMessage: string
  negative?: boolean
  id?: string
}

const InputError: React.FC<InputErrorProps> = ({ className, errorMessage, negative = false, id }) => (
  <div className={classNames('ssb-input-error', negative && 'negative', className)} id={id}>
    <span aria-live='assertive'>{errorMessage}</span>
  </div>
)

export default InputError
