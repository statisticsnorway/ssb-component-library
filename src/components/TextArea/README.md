TextArea
========

> A component for showing TextArea
### Usage


#### HTML

```html
<div class="ssb-text-area ">
    <label for="textarea-input">Label</label>
    <div class="text-area-wrapper">
        <textarea id="textarea-input" placeholder="I can be resized..."></textarea>
    </div>
</div>
```

#### React

```jsx harmony
<TextArea label="Label" placeholder="I can be resized..." handleChange={handleChange} value={someValue} />


```

Available props:

| Name       | Type           | Description  |
| ---------- | ------------- | ----- |
| className   | string | Optional container class|
| disabled | bool | Decides if the textarea is disabled |
| error | bool | Changes design |
| errorMessage | string | Renders an error message underneath input field |
| handleChange | func | Callback function when a value is changed |
| id | String | TextArea Id |
| label | string | Label text |
| negative | bool | Changes design |
| placeholder | string | Placeholder text |
| value | string | TextArea field value |

_Dependencies__
 - [InputError](../InputError)
