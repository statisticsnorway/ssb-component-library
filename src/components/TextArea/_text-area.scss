.ssb-text-area {
  @include roboto;
  cursor: text;
  display: flex;
  flex-direction: column;
  max-width: 400px;
  min-width: 200px;
  position: relative;
  width: 100%;

  &:hover {

    textarea {
      border: 1px solid $ssb-green-4;
      outline: 1px solid $ssb-green-4;

      &:disabled {
        border: 1px solid $ssb-dark-5;
      }
    }
  }

  textarea {
    background: $ssb-white;
    border: 1px solid $ssb-dark-5;
    color: $ssb-dark-6;
    font-size: 16px;
    height: $input-field-height;
    min-height: $input-field-height;
    padding: 8px 12px;
    width: 100%;

    &:focus {
      border: 1px solid $ssb-green-4;
      outline: 1px solid $ssb-green-4;
      outline-offset: 0;
    }
  }

  &.negative {

    .text-area-wrapper {

      ::placeholder {
        color: $ssb-white;
        opacity: 1;
      }

      :-ms-input-placeholder {
        color: $ssb-white;
      }

      ::-ms-input-placeholder {
        color: $ssb-white;
      }
    }

    &:hover {

      textarea {
        border: 1px solid $ssb-green-2;
        outline: 1px solid $ssb-green-2;

        &:disabled {
          border: 1px solid $ssb-white;
        }
      }
    }

    label {
      color: $ssb-white;
    }

    textarea {
      background: transparent;
      border: 1px solid $ssb-white;
      color: $ssb-white;

      &:disabled {
        border: 1px solid $ssb-white;
      }

      &:focus {
        border: 1px solid $ssb-green-2;
        outline: 1px solid $ssb-green-2;
        outline-offset: 0;
      }
    }

    &.error {

      textarea {
        border: 1px solid $ssb-red-3;
        outline: 1px solid $ssb-red-3;
      }
    }
  }

  &.error {

    input {
      border: 1px solid $ssb-red-3;
      outline: 1px solid $ssb-red-3;
    }
  }
}
