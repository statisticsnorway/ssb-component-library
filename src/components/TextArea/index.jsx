import React, { useState } from 'react';
import PropTypes from 'prop-types';
import uuid from 'uuid/v4';
import InputError from '../InputError';

const TextArea = ({
	ariaLabel, className, disabled, error, errorMessage, handleChange, id, label, negative, placeholder, value,
}) => {
	const [inputValue, setValue] = useState(value);
	const inputId = id || uuid();
	const handleInputChange = e => {
		setValue(e.target.value);
		handleChange(e.target.value);
	};

	return (
		<div className={`ssb-text-area${negative ? ' negative' : ''}${error ? ' error' : ''} ${className}`}>
			{label && <label htmlFor={inputId}>{label}</label>}
			<div className="text-area-wrapper">
				<textarea
					id={inputId}
					disabled={disabled}
					value={inputValue}
					onChange={e => handleInputChange(e)}
					placeholder={placeholder}
					aria-label={ariaLabel}
				/>
			</div>
			{error && (errorMessage && (
				<InputError errorMessage={errorMessage} negative={negative} />
			))}
		</div>
	);
};

TextArea.defaultProps = {
	className: '',
	disabled: false,
	error: false,
	handleChange: () => {},
	negative: false,
};

TextArea.propTypes = {
	ariaLabel: PropTypes.string,
	className: PropTypes.string,
	disabled: PropTypes.bool,
	error: PropTypes.bool,
	errorMessage: PropTypes.string,
	handleChange: PropTypes.func,
	id: PropTypes.string,
	label: PropTypes.string,
	negative: PropTypes.bool,
	placeholder: PropTypes.string,
	value: PropTypes.string,
};

export default TextArea;
