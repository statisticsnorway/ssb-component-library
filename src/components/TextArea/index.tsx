import React, { useState, ChangeEvent, FocusEvent } from 'react'
import { useId } from '../../utils/useId'
import InputError from '../InputError'
import classNames from '../../utils/utils'

interface TextAreaProps {
  ariaLabel?: string
  className?: string
  disabled?: boolean
  error?: boolean
  errorMessage?: string
  handleChange?: (value: string) => void
  onFocus?: (e: FocusEvent<HTMLTextAreaElement>) => void
  onBlur?: (e: FocusEvent<HTMLTextAreaElement>) => void
  id?: string
  label?: string
  negative?: boolean
  placeholder?: string
  value?: string
  rows?: number
  cols?: number
}

const TextArea: React.FC<TextAreaProps> = ({
  ariaLabel,
  className,
  disabled = false,
  error = false,
  errorMessage,
  handleChange = () => {},
  id,
  label,
  negative = false,
  placeholder,
  value,
  rows,
  cols,
  onBlur = () => {},
  onFocus = () => {},
}) => {
  const [inputValue, setValue] = useState(value)
  const inputId = id || useId()
  const handleInputChange = (e: ChangeEvent<HTMLTextAreaElement>) => {
    setValue(e.target.value)
    handleChange(e.target.value)
  }

  return (
    <div className={classNames('ssb-text-area', negative && 'negative', error && 'error', className)}>
      {label && <label htmlFor={inputId}>{label}</label>}
      <div className='text-area-wrapper'>
        <textarea
          id={inputId}
          disabled={disabled}
          value={inputValue}
          onChange={(e) => handleInputChange(e)}
          onFocus={onFocus}
          onBlur={onBlur}
          placeholder={placeholder}
          aria-label={ariaLabel}
          rows={rows}
          cols={cols}
          aria-describedby={error && errorMessage ? `error_${inputId}` : undefined}
        />
      </div>
      {error && errorMessage && <InputError errorMessage={errorMessage} negative={negative} id={`error_${inputId}`} />}
    </div>
  )
}

export default TextArea
