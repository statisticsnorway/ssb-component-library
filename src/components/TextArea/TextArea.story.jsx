import React from 'react';
import { storiesOf } from '@storybook/react';
import centered from '@storybook/addon-centered/react';
import TextArea from './index';

let someValue = '';
const handleChange = e => {
	someValue = e;
};

storiesOf('TextArea', module).addDecorator(centered)
	.add('Default', () => (
		<div style={{ width: '500px' }}>
			<TextArea label="Label" placeholder="I can be resized..." handleChange={handleChange} value={someValue} />
		</div>
	))
	.add('Negative', () => (
		<div style={{
			width: '40em',
			height: '20em',
			background: '#274247',
			display: 'flex',
			flexDirection: 'column',
			justifyContent: 'center',
			alignItems: 'center',
		}}
		>
			<TextArea negative label="Label" handleChange={handleChange} value={someValue} />
			<TextArea negative label="Label" handleChange={handleChange} value={someValue} error errorMessage="You made a mistake, fool! What the hell" />
		</div>
	));
