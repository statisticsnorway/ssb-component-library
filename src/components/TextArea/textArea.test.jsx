import React from 'react';
import {shallow} from 'enzyme';
import TextArea from './index';
import InputError from '../InputError';

jest.mock('uuid/v4', () => jest.fn(() => 1));

describe('Input component', () => {
	test('Matches the snapshot', () => {
		const wrapper = shallow(<TextArea>Input</TextArea>);
		expect(wrapper).toMatchSnapshot();
	});

	test('Renders label if added', () => {
		const wrapper = shallow(<TextArea label="Label" />);
		expect(wrapper.find('label').exists()).toEqual(true);
	});

	test('Adds optional classes to wrapper', () => {
		const wrapper = shallow(<TextArea negative error />);
		expect(wrapper.find('.ssb-text-area').hasClass('negative')).toEqual(true);
		expect(wrapper.find('.ssb-text-area').hasClass('error')).toEqual(true);
	});

	test('handleInputChange', () => {
		const handleChange = jest.fn();
		const wrapper = shallow(<TextArea handleChange={handleChange} />);
		wrapper.find('textarea').simulate('change', {
			target: { value: 'hello' }
		});
		expect(handleChange).toBeCalledWith('hello');
	});

	test('Renders an error message on error', () => {
		const wrapper = shallow(<TextArea error errorMessage="An error" />);
		expect(wrapper.find(InputError)).toHaveLength(1);
	})

});
