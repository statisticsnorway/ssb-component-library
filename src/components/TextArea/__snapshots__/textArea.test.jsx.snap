// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component Matches the snapshot 1`] = `
<div
  className="ssb-text-area "
>
  <div
    className="text-area-wrapper"
  >
    <textarea
      disabled={false}
      id={1}
      onChange={[Function]}
    />
  </div>
</div>
`;
